cdc/sinkv2/eventsink/cloudstorage/cloud_storage_dml_sink.go

```diff
@@ -0,0 +1,176 @@
+	encodingWorkers []*encodingWorker
+	// writer is a dmlWriter which manages a group of dmlWorkers and
+	// sends encoded messages to individual dmlWorkers.
+	writer *dmlWriter
+	// tableSeqMap maintains a <versionTable, sequenceNumber> mapping.
+	tableSeqMap map[versionedTable]uint64
+}
+
+// NewCloudStorageSink creates a cloud storage sink.
+func NewCloudStorageSink(ctx context.Context,
+	sinkURI *url.URL,
+	replicaConfig *config.ReplicaConfig,
+	errCh chan error,
+) (*sink, error) {
+	s := &sink{}
+	// create cloud storage config and then apply the params of sinkURI to it.
+	cfg := cloudstorage.NewConfig()
+	err := cfg.Apply(ctx, sinkURI)
+	if err != nil {
+		return nil, err
+	}
+
+	// parse backend storage from sinkURI.
+	bs, err := storage.ParseBackend(sinkURI.String(), &storage.BackendOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// create an external storage.
+	storage, err := storage.New(ctx, bs, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// fetch protocol from replicaConfig defined by changefeed config file.
+	protocol, err := util.GetProtocol(replicaConfig.Sink.Protocol)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// get cloud storage file extension according to the specific protocol.
+	ext := util.GetFileExtension(protocol)
+	encoderConfig, err := util.GetEncoderConfig(sinkURI, protocol, replicaConfig, defaultMaxMessageBytes)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	encoderBuilder, err := builder.NewEventBatchEncoderBuilder(ctx, encoderConfig)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrCloudStorageInvalidConfig, err)
+	}
+
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
+	s.writer = newDMLWriter(ctx, changefeedID, storage, cfg, ext, errCh)
+	s.msgChan = chann.New[eventFragment]()
+	s.tableSeqMap = make(map[versionedTable]uint64)
+
+	// create a group of encoding workers.
+	for i := 0; i < defaultEncodingConcurrency; i++ {
+		encoder := encoderBuilder.Build()
+		w := newEncodingWorker(i+1, changefeedID, encoder, s.writer, errCh)
+		w.run(ctx, s.msgChan)
+		s.encodingWorkers = append(s.encodingWorkers, w)
+	}
+
+	return s, nil
+}
+
+// WriteEvents write events to cloud storage sink.
+func (s *sink) WriteEvents(txns ...*eventsink.CallbackableEvent[*model.SingleTableTxn]) error {
+	var tbl versionedTable
+	for _, txn := range txns {
+		tbl = versionedTable{
+			TableName:    *txn.Event.Table,
+			TableVersion: txn.Event.TableVersion,
+		}
+		s.tableSeqMap[tbl]++
+		// emit a TxnCallbackableEvent encoupled with a sequence number starting from one.
+		s.msgChan.In() <- eventFragment{
+			seqNumber:    s.tableSeqMap[tbl],
+			tableName:    tbl.TableName,
+			tableVersion: tbl.TableVersion,
+			event:        txn,
+		}
+	}
+
+	// emit an eventFragment with nil TxnCallbackableEvent and with current sequence number of the table.
+	// this special eventFragment is used to mark the end of several TxnCallbackableEvent.
+	s.msgChan.In() <- eventFragment{
```

Comment:
Here it's supposed that `txns` come from one table? It's not a documented behavior of `EventSink.WriteEvents`.